### PR TITLE
feat: [1/2] [DeepEP] Fuse shared expert into MoE dispatch under EP

### PR DESCRIPTION
**Pull request by xutizhou**
## Motivation

In DeepSeek V3/R1 with expert parallelism (EP), each rank currently computes the shared expert **separately** from routed experts. This means shared expert is not part of the DeepEP dispatch/compute/combine pipeline — it runs as a standalone forward pass on every rank.

This PR fuses the shared expert into the MoE path by treating it as an additional routed expert (the "9th expert" per rank). Instead of computing shared expert separately, it is dispatched through DeepEP alongside routed experts, assigned to the source rank's local shared expert slot.

This is the **first half** of the waterfill feature split (requested by @ch-wan in #19290):
- **[1/2] This PR**: Fuse shared expert into DeepEP dispatch (fixed local assignment)
- **[2/2] Follow-up PR**: Add waterfill load balancing to dynamically assign shared expert to the least-loaded rank

coauthor: @AichenF

## Modifications

**New file: `python/sglang/srt/layers/moe/deepep_shared_expert_fusion.py`** (65 lines)
- `LOCAL_SHARED_MARKER`: sentinel value for invalid expert ID (DeepEP ignores `expert_id < 0`)
- `expand_topk_with_shared_expert()`: expands topk output `[N, 8] → [N, 9]` with expert ID remapping and shared expert always assigned to source rank

**Modified files:**

- **`models/deepseek_v2.py`** (+138/−22):
  - Adjusts `num_experts` and `top_k` to account for extra shared expert slots (`+ep_size` experts, `+1` topk)
  - Skips standalone shared expert forward when fusion is enabled
  - Calls `expand_topk_with_shared_expert()` after TopK selection to append shared expert column
  - Allows shared expert fusion under EP when `--enable-deepep-waterfill` is set (bypasses existing EP disable checks)

- **`layers/moe/fused_moe_triton/layer.py`** (+40/−2):
  - Expert ID remapping in weight loading: maps checkpoint expert IDs to expanded layout (original routed experts + 1 shared expert slot per rank)
  - Adjusts shared expert detection threshold for the expanded expert ID space

- **`server_args.py`** (+29/−1):
  - Adds `--enable-deepep-waterfill` CLI flag with validation (requires `moe_a2a_backend='deepep'`)
  - Updates `--disable-shared-experts-fusion` help text

## Usage

```bash
python3 -m sglang.launch_server \
    --model-path <DeepSeek-V3-path> \
    --tp 16 --dp-size 16 --nnodes 2 --node-rank {0|1} \
    --dist-init-addr <addr>:<port> \
    --moe-a2a-backend deepep --deepep-mode normal \
    --enable-dp-attention \
    --enable-deepep-waterfill
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 1]
```diff
@@ -570,13 +561,18 @@ def _map_global_expert_id_to_local_expert_id(self, expert_id: int) -> int:
         num_local_routed_experts = (
             self.num_local_experts - self.num_fused_shared_experts
         )
+        # DeepEP fusion: num_experts is expanded by moe_ep_size; adjust to original counts.
+        if self.is_deepep_shared_expert_fusion:
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: `num_local_experts` has been adjusted by `moe_ep_size`. Is this part needed?
https://github.com/xutizhou/sglang/blob/5eedb06110ef71dc3aa82cb89e455e20e594342a/python/sglang/srt/layers/moe/fused_moe_triton/layer.py#L214C14-L216

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 2]
```diff
@@ -622,7 +618,12 @@ def weight_loader(
             )
             return
 
-        if expert_id >= self.num_experts - self.num_fused_shared_experts:
+        # DeepEP fusion: shared experts start at num_experts - moe_ep_size, not num_fused_shared_experts.
+        if expert_id >= self.num_experts - (
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Same question here. I don't understand why we need to specially consider deepep fusion as it's supported for EP + AMD

File: python/sglang/srt/layers/moe/topk.py [hunk 3]
```diff
@@ -776,15 +757,16 @@ def biased_grouped_topk_gpu(
         num_experts // num_expert_group if num_expert_group else num_experts
     )
 
+    # topk for routed experts only (shared experts are appended separately below)
+    topk_routed = topk - num_fused_shared_experts
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/topk.py:798]: Could you confirm if the original implementation of `biased_grouped_topk_gpu` supports shared expert fusion? If yes, why do we need to modify this function?

Such fusion was supported in the pure TP case. Could you explain how topk performs differently under deepep?

File: python/sglang/srt/models/deepseek_v2.py [hunk 4]
```diff
@@ -2134,11 +2225,10 @@ def determine_num_fused_shared_experts(
         elif get_moe_expert_parallel_world_size() > 1 and (
             not _is_hip or torch.cuda.get_device_capability("cuda") < (9, 4)
         ):
-            disable_reason = "Only Deepseek V3/R1 on AMD-platform with capability >= gfx942(MI30x) can use shared experts fusion optimization under expert parallelism."
-        elif disable_reason is None and (
-            get_moe_a2a_backend().is_deepep() or get_moe_a2a_backend().is_mori()
-        ):
-            disable_reason = "Deepseek V3/R1 cannot use shared experts fusion optimization under deepep expert parallelism."
+            disable_reason = (
```
Comment (ch-wan) [inline python/sglang/srt/models/deepseek_v2.py:2260]: We should disable fusion for deepep + TBO / SBO.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 5]
```diff
@@ -195,6 +189,7 @@ def __init__(
         with_bias=False,
         routing_method_type: Optional[RoutingMethodType] = None,
         is_gated: bool = True,
+        is_deepep_shared_expert_fusion: bool = False,
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: This arg is not needed. It is equivalent to deepep + num_fused_shared_experts > 0

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 6]
```diff
@@ -218,6 +213,18 @@ def __init__(
             num_experts - num_fused_shared_experts
         ) // self.moe_ep_size + num_fused_shared_experts
 
+        # Precompute routing parameters shared by AMD and DeepEP fusion paths.
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Comments generated by coding agents. Should be removed.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 7]
```diff
@@ -218,6 +213,18 @@ def __init__(
...
+        # Both reduce to the same structure: _num_global_routed=256, _num_local_routed=16.
+        if is_deepep_shared_expert_fusion:
+            self._num_global_routed = num_experts - self.moe_ep_size
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Should it be related to `num_fused_shared_experts`? For example, deepseek-v2 has 2 shared experts.

File: python/sglang/srt/models/deepseek_v2.py [hunk 8]
```diff
@@ -2113,7 +2193,34 @@ def determine_num_fused_shared_experts(
...
+        server_args = get_global_server_args()
+
+        if server_args.disable_shared_experts_fusion:
+            return
+
+        # SBO/TBO overlap shared expert with A2A dispatch/combine, which is
+        # incompatible with fusing shared expert into the MoE kernel.
+        if is_sbo_enabled() or is_tbo_enabled():
+            server_args.disable_shared_experts_fusion = True
+            log_info_on_rank0(
```
Comment (ch-wan) [inline python/sglang/srt/models/deepseek_v2.py]: Why not setting disable_info?

File: python/sglang/srt/models/deepseek_v2.py [hunk 9]
```diff
@@ -335,6 +337,12 @@ def forward(
         return logits
 
 
+def _is_deepep_class_backend() -> bool:
```
Comment (ch-wan) [inline python/sglang/srt/models/deepseek_v2.py]: This is a general util function. We should not have it in model file.

**Review by ch-wan — DISMISSED** [review]

**Review by ch-wan — COMMENTED** [review]
I accidentally clicked Approve

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py (hunk 7 above)
Comment (xutizhou) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Deepep has to distinguish the shared expert id in different rank for deepep dispatch.

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py (hunk 5 above)
Comment (xutizhou) [reply python/sglang/srt/layers/moe/fused_moe_triton/layer.py]: Replaced with n_shared_experts from model config. Can't fully remove because
  DeepEP passes num_fused_shared_experts=0 to FusedMoE (shared expert handled by
  dispatch, not kernel fusion), so FusedMoE can't infer the layout from existing
  params alone.

**Review by ch-wan — APPROVED** [review]
LGTM
